### PR TITLE
chore(fence): fence config helper py scripts (PXP-4692)

### PR DESCRIPTION
**Pull request by qingyashu**
Little helper `gen3 fence-config` script, replace/delete/extract secrets from config yaml without removing the yaml comments. See `gen3 fence-config` for help. This script is based on Alex's fence config helper script (thanks!!! super helpful). 

Please see doc/fence-config.md for usage. 
Example.1, use `remove-secrets` option to hide secret from a config yaml, for generating public config yaml:

```
gen3 fence-config remove-secrets ./fence-config.yaml pub.yaml
```

Example.2, use `extract-secrets` option to extract secrets from a config yaml, for generating local secret configs: 

```
gen3 fence-config extract-secrets ./fence-config.yaml output.json
```

Example.3, use `merge` option to merge pub and secret configs into one config yaml: 

```
gen3 fence-config merge ./pub-config.yaml ./secret-config.json output.yaml
```

Example.4, use `split` option to split `fence-config.yaml` to public part `pub-config.yaml` and secret  part `secret-config.json`: 

```
gen3 fence-config split ./fence-config.yaml  ./pub-config.yaml ./secret-config.json
```

Please see `gen3/lib/testData/gen3_fence/README.md` for test examples. 

TODO:
- [x] build "merge" command in shell script using "-r" option
- [x] more unit tests using fence configs
- [x] README doc for script
- [x] README doc for config setup
- [ ] move secret template file to fence repo? (will be another PR for fence )
- [x] change fence configuring steps in kube-setup-secrets? 

### Improvements
 - Add fence helper script for extract, delete, and replace configs

### CODE REVIEW

File: doc/fence-config.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Fence config helper
+Little helper to split/merge fence configs, or delete/extract/replace secrets from fence configs
+
+## Use
+
+* `gen3 fence-config extract-secrets [input-yaml-path] [output-json-path]`
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply doc/fence-config.md]: JSON because it's more human readable, and also keeping the yaml without losing comments are also for more human readable. I assume we need extra manual check from devops when splitting the secrets.

File: gen3/lib/testData/gen3_fence/README.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+```
+pub3_level_0: []
+```
+
+*result yaml:*
+
+```
+pub3_level_0:
+```
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply gen3/lib/testData/gen3_fence/README.md]: yeah I think the regex will match for `my-config:` with whatever content after it. I will test this.

File: gen3/lib/fence/fence-secret-config-template.yaml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  GOOGLE_APPLICATION_CREDENTIALS: ___CRED___
+  GOOGLE_STORAGE_CREDS: ___CRED___
+  GOOGLE_ADMIN_EMAIL: ___CRED___
+  GOOGLE_IDENTITY_DOMAIN: ___CRED___
+  GOOGLE_CLOUD_IDENTITY_ADMIN_EMAIL: ___CRED___
+BILLING_PROJECT_FOR_SIGNED_URLS: ___CRED___
+BILLING_PROJECT_FOR_SA_CREDS: ___CRED___
+GUN_MAIL: ___CRED___ 
+EMAIL_SERVER: localhost
+SEND_FROM: example@gmail.com
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply gen3/lib/fence/fence-secret-config-template.yaml]: all those placeholders should be replaced, we only use the key in this template so it doesn't matter what those values are

File: gen3/lib/fence/fence-secret-config-template.yaml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+---
+AWS_CREDENTIALS: ___CRED___
+OPENID_CONNECT: 
+  google: 
+    client_secret: ___CRED___
+  fence: 
+    client_secret: ___CRED___
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply gen3/lib/fence/fence-secret-config-template.yaml]: Alex said it's fine

File: doc/fence-config.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply doc/fence-config.md]: json is more readable for devops to update

File: doc/fence-config.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  Those secrets listed in $GEN3_HOME/gen3/lib/fence/fence-secret-config-template.yaml will be erased.
+
+* `gen3 fence-config remove-secrets-by-template [input-yaml-path] [output-yaml-path] [secret-template-json-path]`
+  
+  Delete secrets and output into a yaml file using customized template. 
+
+* `gen3 fence-config override [input-yaml-path] [output-yaml-path]`
+
+  Override a default fence config. 
+  Default fence config link: https://raw.githubusercontent.com/uc-cdis/fence/master/fence/config-default.yaml.
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply doc/fence-config.md]: we will remove the override command and the link to default config should be in the next command description

File: gen3/bin/fence-config.sh [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  python3 $GEN3_HOME/gen3/lib/fence/config-helper.py -r $inputSecretConfig -c ./config-default.yaml > ./tmp.yaml
+  python3 $GEN3_HOME/gen3/lib/fence/config-helper.py -r $inputPublicConfig -c ./tmp.yaml > $outputPath
+  rm ./config-default.yaml ./tmp.yaml
+  ;;
+"split")
+  curl -s https://raw.githubusercontent.com/uc-cdis/fence/master/fence/config-default.yaml > ./config-default.yaml
+  secretTemplateYaml=$GEN3_HOME/gen3/lib/fence/fence-secret-config-template.yaml
+  inputConfigYaml="$1"
+  outputPublicConfig="$2"
+  outputSecretConfig="$3"
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply gen3/bin/fence-config.sh]: I will add params to this example

File: gen3/lib/fence/config-helper.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+            last_end_position += len(line) + 1 # plus one because "\n"
+            if is_child:
+                last_child_end_position = last_end_position
+    return last_child_end_position
+
+
+def _find_comments_start_position(yaml_config_str, pos):
+    """
+    Find the start position of the comment block in a YAML file string.
+    Assume comments are always above the configurations.
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply gen3/lib/fence/config-helper.py]: this comment would be removed

File: gen3/lib/fence/config-helper.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+    for line in lines:
+        line_without_leading_spaces = line.lstrip(' ')
+        is_comment = (len(line_without_leading_spaces) > 0 and line_without_leading_spaces[0] == "#")
+        leading_spaces_cnt = len(line) - len(line.lstrip(' '))
+        is_child = (leading_spaces_cnt > nested_level * 2)
+        is_empty_line = (len(line_without_leading_spaces) == 0)
+        if not (is_empty_line or is_comment or is_child):
+            break
+        else:
+            last_end_position += len(line) + 1 # plus one because "\n"
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply gen3/lib/fence/config-helper.py]: no it's just 1

File: gen3/lib/fence/config-helper.py [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    )
+    args = parser.parse_args()
+
+    # these three options (-r, -d, -e) are mutually exclusive
+    if args.file_to_replace:
+        print(modify_config_from_files(args.file_to_replace, args.config_file, ConfigAction.REPLACE))
+    if args.file_to_delete:
+        print(modify_config_from_files(args.file_to_delete, args.config_file, ConfigAction.DELETE))
+    if args.template_file:
+        print(extract_config_from_files(args.template_file, args.config_file))
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply gen3/lib/fence/config-helper.py]: Bash script wraps python script and streams the STDOUT into files.

File: gen3/bin/kube-setup-secrets.sh [hunk 10]
```diff
@@ -206,36 +206,48 @@ if [[ -f "$(gen3_secrets_folder)/creds.json" ]]; then # update fence secrets
...
-      echo "job will inject creds.json into fence-config.yaml..."
-      echo "job will also attempt to load old configuration into fence-config.yaml..."
-      echo "NOTE: Some default config values from fence-config.yaml will be replaced"
-      echo "      Run \"gen3 joblogs config-fence\" for details"
-      gen3 job run config-fence CONVERT_OLD_CFG "true"
-
-      # dump fence-config secret into file so user can edit.
-      let count=1
-      while ((count < 50)); do
+      if [[ -f ${fence_config} ]]; then
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply gen3/bin/kube-setup-secrets.sh]: There's a `gen3 fence-config split` I would like to put inside the outer if, so I just keep this indentation. But I will add an error message if the split command doesn't work.

File: gen3/lib/fence/config-helper.py [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+        print((
+            "Error reading {}. Cannot get configuration. Skipping this file. "
+            "Details: {}".format(file_list, str(exc))
+        ))
+        raise
+
+    return configs
+
+
+def _get_all_configs(file_list):
```
[the remark this answers is not in the document]
Comment (qingyashu) [reply gen3/lib/fence/config-helper.py]: will remove this

File: gen3/bin/kube-setup-secrets.sh [hunk 12]
```diff
@@ -206,36 +206,48 @@ if [[ -f "$(gen3_secrets_folder)/creds.json" ]]; then # update fence secrets
   if ! g3kubectl get secrets/fence-config > /dev/null 2>&1; then
     # load updated fence-config.yaml into secret if it exists
     fence_config=$(gen3_secrets_folder)/apis_configs/fence-config.yaml
-    if [[ -f ${fence_config} ]]; then
-      echo "loading fence config from file..."
-      if g3kubectl get secrets/fence-config > /dev/null 2>&1; then
-        g3kubectl delete secret fence-config
-      fi
+    fence_pub_config=$(gen3_secrets_folder)/apis_configs/fence-config-public.yaml
```
Comment (frickjack) [inline gen3/bin/kube-setup-secrets.sh]: The public fence-config.yaml should come from the manifest folder - something like `$(g3k_manifest_folder)/fence/fence-config.yaml`

File: gen3/lib/fence/config-helper.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (qingyashu) [reply gen3/lib/fence/config-helper.py]: I will add a notice to the headers of fence default config file like: "avoid using inline comments because those might be wiped by the cloud automation config helper script). And seems there's only one inline comment in fence default config (S3_BUCKETS/bucket2/region), how about I also move that comment.

File: gen3/bin/kube-setup-secrets.sh [hunk 13]
```diff
@@ -206,36 +206,48 @@ if [[ -f "$(gen3_secrets_folder)/creds.json" ]]; then # update fence secrets
...
+    fence_sec_config=$(gen3_secrets_folder)/apis_configs/fence-config-secret.json
+    if [[ -f ${fence_pub_config} && -f ${fence_sec_config} ]]; then
+      # merge public and secret configs and overwrite ${fence_config}
+      gen3 fence-config merge "${fence_pub_config}"  "${fence_sec_config}" "${fence_config}"
       g3kubectl create secret generic fence-config "--from-file=fence-config.yaml=${fence_config}"
```
Comment (frickjack) [inline gen3/bin/kube-setup-secrets.sh]: The merged fence_config should be a temporary file that gets deleted after the k8s secrets is created.

File: gen3/bin/kube-setup-secrets.sh [hunk 14]
```diff
@@ -206,36 +206,48 @@ if [[ -f "$(gen3_secrets_folder)/creds.json" ]]; then # update fence secrets
...
     # load updated fence-config.yaml into secret if it exists
     fence_config=$(gen3_secrets_folder)/apis_configs/fence-config.yaml
-    if [[ -f ${fence_config} ]]; then
-      echo "loading fence config from file..."
-      if g3kubectl get secrets/fence-config > /dev/null 2>&1; then
-        g3kubectl delete secret fence-config
-      fi
+    fence_pub_config=$(gen3_secrets_folder)/apis_configs/fence-config-public.yaml
+    fence_sec_config=$(gen3_secrets_folder)/apis_configs/fence-config-secret.json
+    if [[ -f ${fence_pub_config} && -f ${fence_sec_config} ]]; then
```
Comment (frickjack) [inline gen3/bin/kube-setup-secrets.sh]: The fence secret config is $(gen3_secrets_folder)/apis_configs/fence-config.yaml - everything under the secrets folder is secret.

The logic should be something like:
* secret part == $(gen3_secrets_fodler)/apis_configs/fence-config.yaml
* public part == $(g3k_manifest_folder)/fence/fence-config-public.yaml if it exists, otherwise empty string
* merge the 2 into a temporary file
* make a secret out of the temporary file
* erase the temporary file
something like that ...
you probably want to put that logic into `bin/secrets.sh`, and remove it from `kube-setup-secrets`

File: gen3/bin/kube-setup-secrets.sh [hunk 15]
```diff
@@ -206,36 +206,48 @@ if [[ -f "$(gen3_secrets_folder)/creds.json" ]]; then # update fence secrets
...
-
-      # dump fence-config secret into file so user can edit.
-      let count=1
-      while ((count < 50)); do
+      if [[ -f ${fence_config} ]]; then
+        echo "loading fence config from file..."
         if g3kubectl get secrets/fence-config > /dev/null 2>&1; then
-          break
+          g3kubectl delete secret fence-config
         fi
```
Comment (frickjack) [inline gen3/bin/kube-setup-secrets.sh]: Be sure to add `/fence/fence-config-public.yaml` files to all the jenkins-* manifest folders under https://github.com/uc-cdis/gitops-qa/ to test this

File: gen3/lib/fence/config-helper.py [hunk 16]
```diff
@@ -69,7 +71,7 @@ def _nested_extract_config(template, source):
     result_content = {}
     try:
         for inner_key, inner_value in template.items():
-            if source is not None and inner_key in source and source[inner_key] is not None:
+            if source.get(inner_key) is not None:
```
Comment (paulineribeyre) [inline gen3/lib/fence/config-helper.py]: sorry, i was wrong in my previous comment, i didn't see you also want to check `source is not None`...
`source.get(inner_key)` does not check for that so you should add it back if it's needed

File: gen3/bin/kube-setup-secrets.sh [hunk 17]
```diff
@@ -440,29 +428,6 @@ if [[ -f "$(gen3_secrets_folder)/creds.json" ]]; then  # update secrets
       PGPASSWORD="$sheepdog_db_password" psql -t -U "$sheepdog_db_user" -h $gdcapi_db_host -d $gdcapi_db_database -c "$sql" || true
     fi
   fi
-
-  # setup the database ...
-  cd "$(gen3_secrets_folder)"
-  if [[ ! -f .rendered_gdcapi_db ]]; then
```
Comment (paulineribeyre) [inline gen3/bin/kube-setup-secrets.sh]: why did you remove this block?

File: doc/fence-config.md [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+  Split a fence config into public yaml config and secret json config. 
+  Just a combination of `gen3 fence-config extract-secrets` and `gen3 fence-config remove-secrets`.
+  Use default fence secret template $GEN3_HOME/gen3/lib/fence/fence-secret-config-template.yaml
+  * `input-config-yaml-path` required: YAML config for input
+  * `output-public-yaml` required: the generated public YAML config path
+  * `output-secret-json` required: the generated secret JSON config path
+
+* `gen3 fence-config replace [input-config-yaml-path] [new-config-yaml-path] [output-yaml-path]`
+
+  Read configs in [input-config-yaml-path] and replace with the configs in [new-config-yaml-path], and output to [output-yaml-path]
```
Comment (paulineribeyre) [inline doc/fence-config.md]:
```suggestion
  Read configs in `input-config-yaml-path` and replace with the configs in `new-config-yaml-path`, and output to `output-yaml-path`
```

File: gen3/lib/fence/config-helper.py [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+            )
+    except AttributeError:
+        # not a dict so replace
+        if value is not None:
+            config_str = _modify_config(action, config_str, traverse_path, value)
+            
+    return config_str
+
+
+def _find_next_sibling_or_parent_position(yaml_config_str, start, nested_level):
```
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply gen3/lib/fence/config-helper.py]: did you check this? if these lists and dicts don't work with the config helper code, please also mention it at the top of the fence config in your fence PR... and if we need to, fix the current config to not use them

File: gen3/bin/kube-setup-secrets.sh (hunk 17 above)
Comment (qingyashu) [reply gen3/bin/kube-setup-secrets.sh]: ooops I don't know what happened. Will add it back.

**Comment by paulineribeyre** [comment]
done in https://github.com/uc-cdis/cloud-automation/pull/1067
